File: SecondTaskJava/src/include/Parser.java
package include;

import java.util.List;

public class Parser {
    private int index = 0;
    private String expression;

    public Parser(String expression) {
        this.expression = expression;
    }

    private Result parsing() {
        return binaryParsing(0);
    }

    private Result binaryParsing(int currentPriority) {
        Result left = expressionsParsing();

        while (true) {
            String op = charParsing();
            int priority = getPriority(op);
            if (priority == 1) {
                if (priority < currentPriority) {
                    index -= op.length();
                    return left;
                }
            } else {
                if (priority <= currentPriority) {
                    index -= op.length();
                    return left;
                }
            }

            Result right = binaryParsing(priority);
            left = new Result(op, left, right);
        }
    }

    private Result expressionsParsing() {
        String currentChar = charParsing();

        if (currentChar.equals("(")) {
            Result result = parsing();
            index++;
            return result;
        }

        if (currentChar.charAt(0) >= 65 && currentChar.charAt(0) <= 90){
            return new Result(currentChar);
        } else {
            return new Result(currentChar, expressionsParsing());
        }
    }

    private String charParsing() {
        if (index == expression.length()) {
            return "";
        }

        if (expression.charAt(index) >= 65 && expression.charAt(index) <= 90) {
            StringBuilder number = new StringBuilder();
            while (index < expression.length() && (expression.charAt(index) >= 65 && expression.charAt(index) <= 90 || expression.charAt(index) >= 48 && expression.charAt(index) <= 57)) {
                number.append(expression.charAt(index++));
            }
            return number.toString();
        }

        String[] operations = {"(", ")", "->", "&", "|", "!"};
        for (String operation : operations) {
            if (operation.charAt(0) == expression.charAt(index)) {
                index += operation.length();
                return operation;
            }
        }
        return "";
    }

    private int getPriority(String operation) {
        if (operation.equals("->")) {
            return 1;
        }
        if (operation.equals("|")) {
            return 2;
        }
        if (operation.equals("&")) {
            return 3;
        }
        if (operation.equals("!")) {
            return 4;
        }
        return 0;
    }

    public Result parse() {
        index = 0;
        return parsing();
    }
}

File: ThirdTaskKotlin/src/include/Parser.kt
package include

class Parser(private val expression: String) {
    private var index = 0
    var is_set = false

    private fun parsing(): Result {
        return binaryParsing(0)
    }

    private fun binaryParsing(current_priority: Int): Result {
        var left = expressionsParsing()

        while (true) {
            val op = charParsing()
            val priority = getPriority(op)
            if (priority == 1) {
                if (priority < current_priority) {
                    index -= op.length
                    return left
                }
            } else {
                if (priority <= current_priority) {
                    index -= op.length
                    return left
                }
            }

            val right = binaryParsing(priority)
            left = Result(op, left, right)
        }
    }

    private fun expressionsParsing(): Result {
        val currentChar = charParsing()

        if (currentChar == "(") {
            val result = parsing()
            index++
            return result
        }

        if (currentChar[0].toInt() in 65..90) {
            if(is_set) {
                set(currentChar)
            }
            return Result(currentChar)
        } else {
            return Result(currentChar, expressionsParsing())
        }

    }

    private fun charParsing(): String {

        if (index == expression.length) {
            return ""
        }

        if (expression[index].toInt() in 65..90) {
            val number = StringBuilder()
            while (index < expression.length && (expression[index].toInt() in 65..90 || expression[index].toInt() in 48..57)) {
                number.append(expression[index++])
            }
            return number.toString()
        }

        val operations = arrayOf("(", ")", "->", "&", "|", "!")
        for (operation in operations) {
            if (operation[0] == expression[index]) {
                index += operation.length
                return operation
            }
        }
        return ""
    }

    private fun getPriority(operation: String): Int {
        if (operation == "->") {
            return 1
        }
        if (operation == "|") {
            return 2
        }
        if (operation == "&") {
            return 3
        }
        if (operation == "!") {
            return 4
        }
        return 0
    }

    fun parse(): Result {
        index = 0
        return parsing()
    }
}